File: Programmers/EnglishEndtalk.swift
import Foundation

// 프로그래머스 영어 끝말잇기, level 2
extension String {
    subscript(idx: Int) -> String? {
        guard (0..<count).contains(idx) else {
            return nil
        }
        
        let target = index(startIndex, offsetBy: idx)
        return String(self[target])
    }
}

func solution(_ n:Int, _ words:[String]) -> [Int] {
    var answer = [0,0]
    var voca : Set = [words[0]]
    
    for i in 1..<words.count {
        guard let last = words[i-1][words[i-1].count-1], let start = words[i][0] else {
            break
        }
        
        if(voca.contains(words[i]) || last != start){
            var num = (i+1) % n
            if (num==0) {
                num = n
            }
            answer[0] = num
            answer[1] = (i/n) + 1
            break
        }
        voca.insert(words[i])
    }
    
    return answer
}

File: Programmers/EnglishEndtalk.cpp


#include <set>
#include <string>
#include <vector>
#include <iostream>

// 프로그래머스 영어 끝말잇기, level 2
using namespace std;

vector<int> solution(int n, vector<string> words) {
    vector<int> answer;
    answer.push_back(0);
    answer.push_back(0);
    
    set<string> voca;
    voca.insert(words[0]);
    
    for(int i=1; i<words.size(); i++){
        char last = words[i-1][words[i-1].size()-1];
        char start = words[i][0];
        
        if(voca.find(words[i]) != voca.end() 
           || last != start){
            answer[0] = (i+1) % n;
            if(answer[0] == 0) answer[0] = n;
            answer[1] = (i/n) + 1;
            break;
        }
        
        voca.insert(words[i]);
    }
    

    return answer;
}
